@nsmith- Nick, perhaps you could give me a help here. I'm testing this patch from a test WMAgent (submit1) and there is one DID that should now have that meta data information.

However, playing interactively with the Rucio client, it looks like the API I'm using is either broken or there is a problem on the Rucio server side, .e.g:
```
>>> res = client.get_did_meta(scope="cms", name="/RelValH125GGgluonfusion_14/Integ_Test-SC_LumiMask_PhEDEx_HG2004_Val_Privv16-v11/NANOAODSIM")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/data/srv/wmagent/v1.3.3/sw/slc7_amd64_gcc630/external/py2-rucio-clients/1.20.5/lib/python2.7/site-packages/rucio/client/didclient.py", line 683, in get_did_meta
    raise exc_cls(exc_msg)
rucio.common.exception.RucioException: An unknown exception occurred.
Details: Feature not in current database
```

--- a/deploy/deploy-wmagent.sh
+++ b/deploy/deploy-wmagent.sh
@@ -340,10 +340,12 @@ sed -i "s+Agent.agentNumber = 0+Agent.agentNumber = $AG_NUM+" $MANAGE_DIR/config
 if [[ "$TEAMNAME" == relval ]]; then
   sed -i "s+config.TaskArchiver.archiveDelayHours = 24+config.TaskArchiver.archiveDelayHours = 336+" $MANAGE_DIR/config.py
   sed -i "s+config.PhEDExInjector.phedexGroup = 'DataOps'+config.PhEDExInjector.phedexGroup = 'RelVal'+" $MANAGE_DIR/config.py
+  sed -i "s+config.RucioInjector.metaDIDProject = 'Production'+config.RucioInjector.metaDIDProject = 'RelVal'+" $MANAGE_DIR/config.py
 elif [[ "$TEAMNAME" == *testbed* ]] || [[ "$TEAMNAME" == *dev* ]]; then
   GLOBAL_DBS_URL=https://cmsweb-testbed.cern.ch/dbs/int/global/DBSReader
   sed -i "s+DBSInterface.globalDBSUrl = 'https://cmsweb.cern.ch/dbs/prod/global/DBSReader'+DBSInterface.globalDBSUrl = '$GLOBAL_DBS_URL'+" $MANAGE_DIR/config.py
   sed -i "s+DBSInterface.DBSUrl = 'https://cmsweb.cern.ch/dbs/prod/global/DBSReader'+DBSInterface.DBSUrl = '$GLOBAL_DBS_URL'+" $MANAGE_DIR/config.py
+  sed -i "s+config.RucioInjector.metaDIDProject = 'Production'+config.RucioInjector.metaDIDProject = 'Test'+" $MANAGE_DIR/config.py
 fi
 
 if [[ "$HOSTNAME" == *fnal.gov ]]; then

--- a/etc/WMAgentConfig.py
+++ b/etc/WMAgentConfig.py
@@ -368,6 +368,7 @@ config.RucioInjector.pollIntervalRules = 43200
 config.RucioInjector.cacheExpiration = 2 * 24 * 60 * 60  # two days
 config.RucioInjector.createBlockRules = True
 config.RucioInjector.RSEPostfix = False  # enable it to append _Test to the RSE names
+config.RucioInjector.metaDIDProject = "Production"
 config.RucioInjector.listTiersToInject = []  # ["NANOAOD", "NANOAODSIM"]
 config.RucioInjector.skipRulesForTiers = ["NANOAOD", "NANOAODSIM"]
 config.RucioInjector.rucioAccount = "OVER_WRITE_BY_SECRETS"

--- a/src/python/WMComponent/RucioInjector/RucioInjectorPoller.py
+++ b/src/python/WMComponent/RucioInjector/RucioInjectorPoller.py
@@ -18,7 +18,7 @@ import time
 from Utils.MemoryCache import MemoryCache
 from Utils.Timers import timeFunction
 from WMCore.DAOFactory import DAOFactory
-from WMCore.Services.Rucio.Rucio import Rucio
+from WMCore.Services.Rucio.Rucio import Rucio, RUCIO_VALID_PROJECT
 from WMCore.WMException import WMException
 from WMCore.WorkerThreads.BaseWorkerThread import BaseWorkerThread
 
@@ -83,6 +83,10 @@ class RucioInjectorPoller(BaseWorkerThread):
         self.createBlockRules = config.RucioInjector.createBlockRules
         self.skipRulesForTiers = config.RucioInjector.skipRulesForTiers
         self.listTiersToInject = config.RucioInjector.listTiersToInject
+        if config.RucioInjector.metaDIDProject not in RUCIO_VALID_PROJECT:
+            msg = "Component configured with an invalid 'project' DID: %s"
+            raise RucioInjectorException(msg % config.RucioInjector.metaDIDProject)
+        self.metaDIDProject = dict(project=config.RucioInjector.metaDIDProject)
 
         # setup cache for container and blocks (containers can be much longer, make 6 days now)
         self.containersCache = MemoryCache(config.RucioInjector.cacheExpiration * 3, set())
@@ -204,7 +208,7 @@ class RucioInjectorPoller(BaseWorkerThread):
             for container in uninjectedData[location]:
                 # same container can be at multiple locations
                 if container not in self.containersCache and container not in newContainers:
-                    if self.rucio.createContainer(container):
+                    if self.rucio.createContainer(container, meta=self.metaDIDProject):
                         logging.info("Container %s inserted into Rucio", container)
                         newContainers.add(container)
                     else:
@@ -227,7 +231,7 @@ class RucioInjectorPoller(BaseWorkerThread):
             for container in uninjectedData[location]:
                 for block in uninjectedData[location][container]:
                     if block not in self.blocksCache:
-                        if self.rucio.createBlock(block, rse=rseName):
+                        if self.rucio.createBlock(block, rse=rseName, meta=self.metaDIDProject):
                             logging.info("Block %s inserted into Rucio", block)
                             newBlocks.add(block)
                         else:

--- a/src/python/WMCore/Services/Rucio/Rucio.py
+++ b/src/python/WMCore/Services/Rucio/Rucio.py
@@ -16,12 +16,33 @@ from rucio.common.exception import (AccountNotFound, DataIdentifierNotFound, Acc
 from WMCore.WMException import WMException
 
 
+RUCIO_VALID_PROJECT = ("Production", "RelVal", "Tier0", "Test", "User")
+
+
 class WMRucioException(WMException):
     """
     _WMRucioException_
     Generic WMCore exception for Rucio
     """
     pass
+
+
+def validateMetaData(did, metaDict, logger):
+    """
+    This function can be extended in the future, for now it will only
+    validate the DID creation metadata, more specifically only the
+    "project" parameter
+    :param did: the DID that will be inserted
+    :param metaDict: a dictionary with all the DID metadata data to be inserted
+    :param logger: a logger object
+    :return: False if validation fails, otherwise True
+    """
+    if metaDict.get("project", "Production") in RUCIO_VALID_PROJECT:
+        return True
+    msg = "DID: %s has an invalid 'project' meta-data value: %s" % (did, metaDict['project'])
+    msg += "The supported 'project' values are: %s" % str(RUCIO_VALID_PROJECT)
+    logger.error(msg)
+    return False
 
 
 class Rucio(object):
@@ -258,6 +279,8 @@ class Rucio(object):
         :return: a boolean to represent whether it succeeded or not
         """
         response = False
+        if not validateMetaData(name, kwargs.get("meta", {}), logger=self.logger):
+            return response
         try:
             # add_container(scope, name, statuses=None, meta=None, rules=None, lifetime=None)
             response = self.cli.add_container(scope, name, **kwargs)
@@ -292,6 +315,8 @@ class Rucio(object):
               being produced
         """
         response = False
+        if not validateMetaData(name, kwargs.get("meta", {}), logger=self.logger):
+            return response
         try:
             # add_dataset(scope, name, statuses=None, meta=None, rules=None, lifetime=None, files=None, rse=None)
             response = self.cli.add_dataset(scope, name, **kwargs)

--- a/test/python/WMCore_t/Services_t/Rucio_t/Rucio_t.py
+++ b/test/python/WMCore_t/Services_t/Rucio_t/Rucio_t.py
@@ -8,7 +8,7 @@ import os
 
 from rucio.client import Client as testClient
 
-from WMCore.Services.Rucio.Rucio import Rucio
+from WMCore.Services.Rucio.Rucio import Rucio, validateMetaData, RUCIO_VALID_PROJECT
 from WMQuality.Emulators.EmulatedUnitTestCase import EmulatedUnitTestCase
 
 DSET = "/SingleElectron/Run2017F-17Nov2017-v1/MINIAOD"
@@ -234,3 +234,19 @@ class RucioTest(EmulatedUnitTestCase):
         # Properly formatted rule, rule manually created
         res = self.myRucio.getRule("1d6ea1d916d5492e81b1bb30ed4aebc1")
         self.assertTrue(res)
+
+    def testMetaDataValidation(self):
+        """
+        Test the `validateMetaData` validation function
+        """
+        for thisProj in RUCIO_VALID_PROJECT:
+            response = validateMetaData("any_DID_name", dict(project=thisProj), self.myRucio.logger)
+            self.assertTrue(response)
+
+        # test with no "project" meta data at all
+        response = validateMetaData("any_DID_name", dict(), self.myRucio.logger)
+        self.assertTrue(response)
+
+        # now an invalid "project" meta data
+        response = validateMetaData("any_DID_name", dict(project="mistake"), self.myRucio.logger)
+        self.assertFalse(response)

